specs/_features/eip7594/das-core.md

```diff
@@ -0,0 +1,262 @@
+  - [Helper functions](#helper-functions)
+    - [`get_custody_columns`](#get_custody_columns)
+    - [`compute_extended_data`](#compute_extended_data)
+    - [`compute_extended_matrix`](#compute_extended_matrix)
+    - [`get_data_column_sidecars`](#get_data_column_sidecars)
+- [Custody](#custody)
+  - [Custody requirement](#custody-requirement)
+  - [Public, deterministic selection](#public-deterministic-selection)
+- [Peer discovery](#peer-discovery)
+- [Extended data](#extended-data)
+- [Column gossip](#column-gossip)
+  - [Parameters](#parameters)
+  - [Reconstruction and cross-seeding](#reconstruction-and-cross-seeding)
+- [Peer sampling](#peer-sampling)
+- [Peer scoring](#peer-scoring)
+- [DAS providers](#das-providers)
+- [A note on fork choice](#a-note-on-fork-choice)
+- [FAQs](#faqs)
+  - [Row (blob) custody](#row-blob-custody)
+  - [Subnet stability](#subnet-stability)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Custom types
+
+We define the following Python custom types for type hinting and readability:
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `DataColumn` | `List[Cell, MAX_BLOBS_PER_BLOCK]` | The data of each column in EIP-7594 |
+| `ExtendedMatrix` | `List[Cell, MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The full data of one-dimensional erasure coding extended blobs (in row major format) |
+| `FlatExtendedMatrix` | `List[BLSFieldElement, MAX_BLOBS_PER_BLOCK * FIELD_ELEMENTS_PER_BLOB * NUMBER_OF_COLUMNS]` | The flattened format of `ExtendedMatrix` |
+
+## Configuration
+
+### Data size
+
+| Name | Value | Description |
+| - | - | - |
+| `NUMBER_OF_COLUMNS` | `uint64((FIELD_ELEMENTS_PER_BLOB * 2) // FIELD_ELEMENTS_PER_CELL)` (= 128) | Number of columns in the extended data matrix. |
+
+### Networking
+
+| Name | Value | Description |
+| - | - | - |
+| `DATA_COLUMN_SIDECAR_SUBNET_COUNT` | `32` | The number of data column sidecar subnets used in the gossipsub protocol |
+
+### Custody setting
+
+| Name | Value | Description |
+| - | - | - |
+| `SAMPLES_PER_SLOT` | `8` | Number of `DataColumn` random samples a node queries per slot |
+| `CUSTODY_REQUIREMENT` | `1` | Minimum number of subnets an honest node custodies and serves samples from |
+| `TARGET_NUMBER_OF_PEERS` | `70` | Suggested minimum peer count |
+
+### Containers
+
+#### `DataColumnSidecar`
+
+```python
+class DataColumnSidecar(Container):
+    index: ColumnIndex  # Index of column in extended matrix
+    column: DataColumn
+    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
+    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
+    signed_block_header: SignedBeaconBlockHeader
+    kzg_commitments_inclusion_proof: Vector[Bytes32, KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH]
+```
+
+### Helper functions
+
+#### `get_custody_columns`
+
+```python
+def get_custody_columns(node_id: NodeID, custody_subnet_count: uint64) -> Sequence[ColumnIndex]:
```

Comment:
We currently have a function that maps node-id -> subnets and I think we plan to also do node-id -> sync-committees etc. 
@Nashatyrev had an idea called Network Sharding, which just generalizes this concept. 

The idea is we just have a single function that maps node-id -> shard.  
Then we assign attestation subnets, sync-committees and now data columns to shards. 

The benefit here is:
- We have a single function that does the mapping and we don't have to make three different forms of it, making the implementation easier and making it easier to debug
- If we implement it like in: https://github.com/ethereum/consensus-specs/pull/3545, then we make it faster for us to discover peers on columns/subnets/sync committees
- The specs I think might simplify, because this function would just be a mapping from column -> shard.

If others agree, I can draft a spec PR that implements this idea, perhaps it would be more explicit that way.

Reply:
If there are 64 shards but we want to split the blob data in 32 parts, do we map each of the 32 parts to two shards?

Reply:
Yep. The number of shards would be the maximum granularity of splitting nodes. Like we could have 128 shards, in case we wanted to divide the network into 128 and assign subnets etc. I think 64 makes sense at the moment tho, because we currently subscribe to two subnets per node, so 2 shards there also. 

However if we want to have 128 data columns and split this up equally, we might want to have 128 shards, then subnets would be mapped to 4 shards (or vice versa)

Reply:
> The idea is we just have a single function that maps node-id -> shard

Can it be extended to consider `custody_subnet_count`? i.e., to map node-id to multiple shards.

> assign attestation subnets, sync-committees and now data columns to shards

Would it separate the long-lived subscriptions and the mid-term-lived subscriptions? We may add a [PeerDAS assignment period](https://github.com/ethereum/consensus-specs/blob/87e9702ab0267b4c2320aff31f09a550f744652b/specs/_features/eip7594/das-core.md#subnet-stability) into the deterministic function.

> I can draft a spec PR that implements this idea, perhaps it would be more explicit that way.

I think it would be good to see a PR for it. One minor suggestion: I like the "Network Sharding" naming. However, it might be ambiguous with the data sharding we're mostly talking about.

Reply:
For reference, the ethresearch post of network sharding is here https://ethresear.ch/t/network-shards-concept-idea/17417

Reply:
> @Nashatyrev had an idea called Network Sharding, which just generalizes this concept.

Here is the concept sketch: https://ethresear.ch/t/network-shards-concept-idea/17417

Reply:
> If others agree, I can draft a spec PR that implements this idea, perhaps it would be more explicit that way.

@AgeManning please feel free to ping me if you want my assistance or maybe want to delegate that to me

Reply:
> Can it be extended to consider custody_subnet_count? i.e., to map node-id to multiple shards.

yep that should be easy to do (If I understand your question). 

> Would it separate the long-lived subscriptions and the mid-term-lived subscriptions? We may add a [PeerDAS assignment period](https://github.com/ethereum/consensus-specs/blob/87e9702ab0267b4c2320aff31f09a550f744652b/specs/_features/eip7594/das-core.md#subnet-stability) into the deterministic function.

This is a good point I'd need to think about. The period of rotation might need to be made a parameter of the function. This would mean that nodes will be on different shards based on this parameter. I guess we shouldn't force rotation periods for different use-cases to be the same. 
In this case, I guess the benefit still is that we have a single function that handles the mappings in a discovery-efficient way.

> @AgeManning please feel free to ping me if you want my assistance or maybe want to delegate that to me

Thanks. I have been traveling and have a conference next week, but I will endeavor to get a PR up for this that we can work from. Feel free to make one earlier if you have the time tho.

Reply:
Apologies for the delay. Here is the proposal: https://github.com/ethereum/consensus-specs/pull/3623

Reply:
> Here is the proposal: #3623

Great! Thanks @AgeManning 

If this PR is adopted then I'd suggest to make `DATA_COLUMN_SIDECAR_SUBNET_COUNT` equal to `NUMBER_OF_COLUMNS`. I.e. one subnet per one column. (that should potentially work well even if number of columns (and thus number of subnets) is of order 2048)

The above would allow push sampling approach: a sampling node could be connected to all (or a subset) of network shards (without subscribing to any column subnets). Some short period (sub-slot) prior to sampling a node subscribes to selected subnets and immediately unsubscribes upon sample receiving. This push approach would allow to 
- avoid extra throughput for sampling: node receives only required samples (columns) 
- almost preserves _unlinkability of queries_: the gossip `subscribe` request could be issued as shortly before expected sample as a second or two 
- a node may always fall back to pull method (req/resp `data_column_sidecars_by_root`) if fails to get a sample via push